Use vector for small dict.

methane opened the pull request.

methane committed:
Implement vector based implementation for small dict.

methane committed:
Use SSE2

methane committed:
Use 8 entries

methane committed:
Use 16-wide vector when SSE2 is available

methane committed:
fix estimate_log2_keysize

lpereira commented on Objects/dictobject.c:
```diff
@@ -377,6 +394,79 @@ dictkeys_set_index(PyDictKeysObject *keys, Py_ssize_t i, Py_ssize_t ix)
+// https://graphics.stanford.edu/~seander/bithacks.html
+
+const uint64_t lsb = 0x0101010101010101ull;
+const uint64_t msb = 0x8080808080808080ull;
```
These should be `static`, otherwise they're going to be exported in the shared library.

methane replied:
thanks.

joernheissler committed:
[docs] Correct typos in SSLContext.sni_callback (GH-30623)

Co-authored-by: Jörn Heissler <nosuchaddress@joern-heissler.de>

sobolevn committed:
bpo-46571: improve `typing.no_type_check` to skip foreign objects (GH-31042)

There are several changes:
1. We now don't explicitly check for any base / sub types, because new name check covers it
2. I've also checked that `no_type_check` do not modify foreign functions. It was the same as with `type`s
3. I've also covered `except TypeError` in `no_type_check` with a simple test case, it was not covered at all
4. I also felt like adding `lambda` test is a good idea: because `lambda` is a bit of both in class bodies: a function and an assignment

<!-- issue-number: [bpo-46571](https://bugs.python.org/issue46571) -->
https://bugs.python.org/issue46571
<!-- /issue-number -->

sobolevn committed:
bpo-46603: improve coverage of `typing._strip_annotations` (GH-31063)

sobolevn committed:
bpo-46685: improve test coverage of `Self` and `Never` in `typing` (GH-31222)

bruchar1 committed:
Counter doc mentions three methods, but lists four (GH-30706)

Was probably caused by the addition of the `total()` method

methane committed:
dict: Use DK_LOG_SIZE in hot loop. (GH-31405)

DK_LOG_SIZE(key) < 8 is faster than DK_SIZE(key) <= 0xff, at least on GCC.
